File: rotate_image.js
//You are given an n x n 2D matrix representing an image, rotate the image by 90 degrees(clockwise).
//
//You have to rotate the image in -place, which means you have to modify the input 2D matrix directly.DO NOT allocate another 2D matrix and do the rotation.

function rotateMatrix(matrix) {
    let n = matrix.length
    let element;

    var result = createMatrix(n, n);

    for (let i = 0; i < n; i++) {
        var clm = (n - i - 1);
        for (let j = 0; j < n; j++) {
            element = matrix[i][j];
            result[j][clm] = element;            
        }
    }

    matrix = result;
    return matrix;
};

function createMatrix(lns, clms) {
    var matrix = [];
    var sub = [];
    
    for (let j = 0; j < clms; j++) {
        sub.push(0);
    }

    for (let i = 0; i < lns; i++) {
        matrix[i] = sub;
    }

    return matrix;
}

function printMatrix(matrix) {
    for (let i = 0; i <= matrix.length; i++) {
        for (let j = 0; i<= matrix.length; j++) {
            console.log(matrix[i][j] + " ");
        }
        console.log("\n");
    }
}

File: RotateImage.java
//You are given an n x n 2D matrix representing an image, rotate the image by 90 degrees(clockwise).
//
//You have to rotate the image in -place, which means you have to modify the input 2D matrix directly.DO NOT allocate another 2D matrix and do the rotation.

class RotateImage {
    public static void main(String[] args) {
        int[][] matrix = {{5,1,9,11},{2,4,8,10},{13,3,6,7},{15,14,12,16}};

        rotate(matrix);
    }
    
    public static void rotate(int[][] matrix) {
        int n = matrix.length;
        int temp;

        for (int i = 0; i < Math.floor(n / 2); i++) {
            for (int j = i; j < n - 1 - i; j++) {
                temp = matrix[i][j];
                matrix[i][j] = matrix[n - 1 - j][i];
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j];
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i];
                matrix[j][n - 1 - i] = temp;
            }
        }
    }
}